keyboards/edgepromx/matrix.c

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2023 Jernej Turnsek <jernej.turnsek@gmail.com>
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#include "matrix.h"
+#include <string.h>
+#include "timer.h"
+#include "wait.h"
+#include "debug.h"
+#include "i2c_master.h"
+#include "edgepromx.h"
+
+extern i2c_status_t mcp23017_status;
+
+#define I2C_WRITE   0x00
+#define I2C_READ    0x01
+
+// For a better understanding of the i2c protocol, this is a good read:
+// https://www.robot-electronics.co.uk/i2c-tutorial
+
+// I2C address:
+// See the datasheet, section 3.3.1 on addressing I2C devices and figure 3-6 for an
+// illustration
+// http://ww1.microchip.com/downloads/en/devicedoc/20001952c.pdf
+// All address pins of the mcp23017 are connected to the ground on the ferris
+// | 0  | 1  | 0  | 0  | A2 | A1 | A0 |
+// | 0  | 1  | 0  | 0  | 0  | 0  | 0  |
+#define I2C_ADDR 0b0100000
+#define I2C_ADDR_WRITE ((I2C_ADDR << 1) | I2C_WRITE)
+#define I2C_ADDR_READ ((I2C_ADDR << 1) | I2C_READ)
+
+// Register addresses
+#define IODIRA 0x00  // i/o direction register
+#define IODIRB 0x01
+#define GPPUA 0x0C  // GPIO pull-up resistor register
+#define GPPUB 0x0D
+#define GPIOA 0x12  // general purpose i/o port register (write modifies OLAT)
+#define GPIOB 0x13
+#define OLATA 0x14  // output latch register
+#define OLATB 0x15
+
+bool         i2c_initialized = 0;
+i2c_status_t mcp23017_status = I2C_ADDR;
+
+uint8_t init_mcp23017(void) {
```

Comment:
Could this be converted to use the `driver/gpio/mcp23018.c` code?

Reply:
I am actually using mcp23017 chip. Although it could be done, but for now I would like to stay with my current implementation.

Reply:
Just a heads up, the mcp23017 and mcp23018 use the same registers and config.  They're interchangeable.